File: frontend/src/services/appointmentService.ts
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:5000/api';

const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add auth token to requests
api.interceptors.request.use((config) => {
  const token = localStorage.getItem('token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

// Appointment Types
export interface CreateAppointmentDto {
  salonId: string;
  serviceId: string;
  staffId: string;
  startTime: string; // ISO date string
  notes?: string;
  customerNotes?: string;
}

export interface AppointmentDto {
  id: string;
  tenantId: string;
  salonId: string;
  salonName: string;
  customerId: string;
  customerName: string;
  customerEmail: string;
  customerPhone?: string;
  staffId: string;
  staffName: string;
  serviceId: string;
  serviceName: string;
  serviceDuration: number;
  servicePrice: number;
  startTime: string;
  endTime: string;
  status: AppointmentStatus;
  notes?: string;
  customerNotes?: string;
  staffNotes?: string;
  cancellationReason?: string;
  cancelledAt?: string;
  totalPrice: number;
  depositPaid: number;
  reminderSent: boolean;
  createdAt: string;
  // Computed fields
  duration?: number; // serviceDuration alias for compatibility
}

export interface AppointmentDetailsDto extends AppointmentDto {
  salon: SalonDto;
  service: ServiceDto;
  staff: StaffDto;
  payments: PaymentDto[];
}

export interface SalonDto {
  id: string;
  tenantId: string;
  name: string;
  address: string;
  city: string;
  phone: string;
  email: string;
  state?: string;
  postalCode?: string;
  latitude?: number;
  longitude?: number;
  businessHours?: string;
  isActive: boolean;
  averageRating: number;
  reviewCount: number;
  createdAt: string;
}

export interface ServiceDto {
  id: string;
  name: string;
  description?: string;
  durationMinutes: number;
  price: number;
  imageUrl?: string;
  categoryId?: string;
  categoryName?: string;
  isActive: boolean;
}

export interface StaffDto {
  id: string;
  tenantId: string;
  salonId: string;
  firstName: string;
  lastName: string;
  fullName: string;
  email: string;
  phone?: string;
  bio?: string;
  photoUrl?: string;
  profilePictureUrl?: string;
  averageRating: number;
  status: string;
  invitationStatus: string;
  roleId: string;
  roleName?: string;
}

export interface PaymentDto {
  id: string;
  amount: number;
  method: string;
  status: string;
}

export interface AvailableTimeSlotDto {
  date: string;
  startTime: string; // TimeSpan format
  endTime: string;
  durationMinutes: number;
}

export interface RescheduleAppointmentDto {
  newStartTime: string;
  newStaffId?: string;
}

export interface CancelAppointmentDto {
  cancellationReason?: string;
}

export const AppointmentStatus = {
  Pending: 0,
  Confirmed: 1,
  CheckedIn: 2,
  InProgress: 3,
  Completed: 4,
  Cancelled: 5,
  NoShow: 6,
} as const;

export type AppointmentStatus = typeof AppointmentStatus[keyof typeof AppointmentStatus];

// Legacy types for backward compatibility with dashboard
export interface Appointment {
  id: string;
  customerName: string;
  customerPhone: string;
  customerEmail: string;
  serviceName: string;
  servicePrice: number;
  staffName: string;
  appointmentDate: string;
  appointmentTime: string;
  status: 'scheduled' | 'completed' | 'cancelled' | 'in-progress';
  duration: number;
  notes?: string;
}

export interface Service {
  id: string;
  name: string;
  description: string;
  price: number;
  duration: number;
  category: string;
  isActive: boolean;
}

export interface Staff {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  specialties: string[];
  isActive: boolean;
  avatar?: string;
}

export interface Customer {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  lastVisit?: string;
  totalAppointments: number;
  totalSpent: number;
}

export interface DashboardStats {
  todayAppointments: number;
  totalCustomers: number;
  monthlyRevenue: number;
  averageRating: number;
  appointmentGrowth: number;
  customerGrowth: number;
  revenueGrowth: number;
  ratingGrowth: number;
}

// Helper function to convert AppointmentDto to legacy Appointment format
const convertToLegacyAppointment = (dto: AppointmentDto): Appointment => {
  const startTime = new Date(dto.startTime);
  const statusMap: Record<number, 'scheduled' | 'completed' | 'cancelled' | 'in-progress'> = {
    0: 'scheduled', // Pending
    1: 'scheduled', // Confirmed
    2: 'in-progress', // CheckedIn
    3: 'in-progress', // InProgress
    4: 'completed', // Completed
    5: 'cancelled', // Cancelled
    6: 'cancelled', // NoShow
  };

  return {
    id: dto.id,
    customerName: dto.customerName,
    customerPhone: dto.customerPhone || '',
    customerEmail: dto.customerEmail,
    serviceName: dto.serviceName,
    servicePrice: dto.servicePrice,
    staffName: dto.staffName,
    appointmentDate: startTime.toISOString().split('T')[0],
    appointmentTime: startTime.toLocaleTimeString('tr-TR', { hour: '2-digit', minute: '2-digit' }),
    status: statusMap[dto.status] || 'scheduled',
    duration: dto.serviceDuration,
    notes: dto.notes,
  };
};

// API Service
export const appointmentService = {
  // Create appointment
  createAppointment: async (data: CreateAppointmentDto): Promise<AppointmentDto> => {
    const response = await api.post('/appointments', data);
    return response.data;
  },

  // Get appointment details
  getAppointmentDetails: async (id: string): Promise<AppointmentDetailsDto> => {
    const response = await api.get(`/appointments/${id}`);
    return response.data;
  },

  // Get customer's appointments
  getMyAppointments: async (
    startDate?: string,
    endDate?: string,
    status?: AppointmentStatus
  ): Promise<AppointmentDto[]> => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);
    if (status !== undefined) params.append('status', status.toString());

    const response = await api.get(`/appointments/my?${params.toString()}`);
    return response.data;
  },

  // Get staff appointments
  getStaffAppointments: async (
    staffId: string,
    startDate?: string,
    endDate?: string
  ): Promise<AppointmentDto[]> => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);

    const response = await api.get(`/appointments/staff/${staffId}?${params.toString()}`);
    return response.data;
  },

  // Get salon appointments
  getSalonAppointments: async (
    salonId: string,
    startDate?: string,
    endDate?: string
  ): Promise<AppointmentDto[]> => {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);

    const response = await api.get(`/appointments/salon/${salonId}?${params.toString()}`);
    return response.data;
  },

  // Update appointment status
  updateAppointmentStatus: async (
    id: string,
    status: AppointmentStatus,
    cancellationReason?: string
  ): Promise<AppointmentDto> => {
    const response = await api.put(`/appointments/${id}/status`, {
      status,
      cancellationReason,
    });
    return response.data;
  },

  // Reschedule appointment
  rescheduleAppointment: async (
    id: string,
    data: RescheduleAppointmentDto
  ): Promise<AppointmentDto> => {
    const response = await api.put(`/appointments/${id}/reschedule`, data);
    return response.data;
  },

  // Cancel appointment
  cancelAppointment: async (id: string, cancellationReason?: string): Promise<void> => {
    await api.delete(`/appointments/${id}`, {
      data: { cancellationReason },
    });
  },

  // Get available time slots
  getAvailableTimeSlots: async (
    staffId: string,
    date: string,
    serviceDurationMinutes: number = 60
  ): Promise<AvailableTimeSlotDto[]> => {
    const params = new URLSearchParams({
      staffId,
      date,
      serviceDurationMinutes: serviceDurationMinutes.toString(),
    });

    const response = await api.get(`/appointments/availability?${params.toString()}`);
    return response.data;
  },

  // Get available staff for a service at a specific time
  getAvailableStaff: async (
    salonId: string,
    serviceId: string,
    dateTime: string,
    durationMinutes: number = 60
  ): Promise<StaffDto[]> => {
    const params = new URLSearchParams({
      salonId,
      serviceId,
      dateTime,
      durationMinutes: durationMinutes.toString(),
    });

    const response = await api.get(`/appointments/availability/staff?${params.toString()}`);
    return response.data;
  },

  // Get salon-wide availability
  getSalonAvailability: async (
    salonId: string,
    serviceId: string,
    date: string
  ): Promise<Record<string, AvailableTimeSlotDto[]>> => {
    const params = new URLSearchParams({
      serviceId,
      date,
    });

    const response = await api.get(`/appointments/availability/salon/${salonId}?${params.toString()}`);
    return response.data;
  },

  // Legacy methods for dashboard compatibility
  getAllAppointments: async (): Promise<Appointment[]> => {
    const response = await api.get('/appointments/my');
    const appointments: AppointmentDto[] = response.data;
    return appointments.map(convertToLegacyAppointment);
  },

  getTodayAppointments: async (): Promise<Appointment[]> => {
    const today = new Date().toISOString().split('T')[0];
    const response = await api.get(`/appointments/my?startDate=${today}&endDate=${today}`);
    const appointments: AppointmentDto[] = response.data;
    return appointments.map(convertToLegacyAppointment);
  },

  getDailyAppointments: async (date: string): Promise<Appointment[]> => {
    const response = await api.get(`/appointments/my?startDate=${date}&endDate=${date}`);
    const appointments: AppointmentDto[] = response.data;
    return appointments.map(convertToLegacyAppointment);
  },

  getWeeklyAppointments: async (startDate: string): Promise<Appointment[]> => {
    const start = new Date(startDate);
    const end = new Date(start);
    end.setDate(start.getDate() + 6);
    const endDate = end.toISOString().split('T')[0];

    const response = await api.get(`/appointments/my?startDate=${startDate}&endDate=${endDate}`);
    const appointments: AppointmentDto[] = response.data;
    return appointments.map(convertToLegacyAppointment).sort((a, b) => {
      if (a.appointmentDate === b.appointmentDate) {
        return a.appointmentTime.localeCompare(b.appointmentTime);
      }
      return a.appointmentDate.localeCompare(b.appointmentDate);
    });
  },

  getMonthlyAppointments: async (year: number, month: number): Promise<Appointment[]> => {
    const startDate = new Date(year, month, 1).toISOString().split('T')[0];
    const endDate = new Date(year, month + 1, 0).toISOString().split('T')[0];

    const response = await api.get(`/appointments/my?startDate=${startDate}&endDate=${endDate}`);
    const appointments: AppointmentDto[] = response.data;
    return appointments.map(convertToLegacyAppointment).sort((a, b) => {
      if (a.appointmentDate === b.appointmentDate) {
        return a.appointmentTime.localeCompare(b.appointmentTime);
      }
      return a.appointmentDate.localeCompare(b.appointmentDate);
    });
  },

  getAppointmentsByDateRange: async (startDate: string, endDate: string): Promise<Appointment[]> => {
    const response = await api.get(`/appointments/my?startDate=${startDate}&endDate=${endDate}`);
    const appointments: AppointmentDto[] = response.data;
    return appointments.map(convertToLegacyAppointment).sort((a, b) => {
      if (a.appointmentDate === b.appointmentDate) {
        return a.appointmentTime.localeCompare(b.appointmentTime);
      }
      return a.appointmentDate.localeCompare(b.appointmentDate);
    });
  },

  // Placeholder methods - these need proper implementation
  getAllServices: async (): Promise<Service[]> => {
    // TODO: Implement with real service API
    return [];
  },

  createService: async (service: Omit<Service, 'id'>): Promise<Service> => {
    // TODO: Implement with real service API
    return { ...service, id: '' };
  },

  getAllStaff: async (): Promise<Staff[]> => {
    // TODO: Implement with real staff API
    return [];
  },

  createStaff: async (staff: Omit<Staff, 'id'>): Promise<Staff> => {
    // TODO: Implement with real staff API
    return { ...staff, id: '' };
  },

  getAllCustomers: async (): Promise<Customer[]> => {
    // TODO: Implement with real customer API
    return [];
  },

  getDashboardStats: async (): Promise<DashboardStats> => {
    // TODO: Implement with real dashboard API
    return {
      todayAppointments: 0,
      totalCustomers: 0,
      monthlyRevenue: 0,
      averageRating: 0,
      appointmentGrowth: 0,
      customerGrowth: 0,
      revenueGrowth: 0,
      ratingGrowth: 0,
    };
  },
};

export default appointmentService;

File: src/RendevumVar.Application/Services/AppointmentService.cs
using Microsoft.Extensions.Logging;
using RendevumVar.Application.DTOs;
using RendevumVar.Application.Interfaces;
using RendevumVar.Core.Entities;
using RendevumVar.Core.Enums;
using RendevumVar.Core.Repositories;

namespace RendevumVar.Application.Services;

public class AppointmentService : IAppointmentService
{
    private readonly IAppointmentRepository _appointmentRepository;
    private readonly IServiceRepository _serviceRepository;
    private readonly IStaffRepository _staffRepository;
    private readonly ISalonRepository _salonRepository;
    private readonly INotificationService _notificationService;
    private readonly ILogger<AppointmentService> _logger;

    public AppointmentService(
        IAppointmentRepository appointmentRepository,
        IServiceRepository serviceRepository,
        IStaffRepository staffRepository,
        ISalonRepository salonRepository,
        INotificationService notificationService,
        ILogger<AppointmentService> logger)
    {
        _appointmentRepository = appointmentRepository;
        _serviceRepository = serviceRepository;
        _staffRepository = staffRepository;
        _salonRepository = salonRepository;
        _notificationService = notificationService;
        _logger = logger;
    }

    public async Task<AppointmentDto> CreateAppointmentAsync(
        CreateAppointmentDto dto,
        Guid customerId,
        Guid tenantId)
    {
        // Validate service exists
        var service = await _serviceRepository.GetByIdAsync(dto.ServiceId);
        if (service == null || service.TenantId != tenantId)
            throw new InvalidOperationException("Service not found");

        // Validate staff exists and belongs to salon
        var staff = await _staffRepository.GetByIdAsync(dto.StaffId);
        if (staff == null || staff.TenantId != tenantId || staff.SalonId != dto.SalonId)
            throw new InvalidOperationException("Staff not found or does not work at this salon");

        // Validate salon exists
        var salon = await _salonRepository.GetByIdAsync(dto.SalonId);
        if (salon == null || salon.TenantId != tenantId)
            throw new InvalidOperationException("Salon not found");

        // Calculate end time
        var endTime = dto.StartTime.AddMinutes(service.DurationMinutes);

        // Check for conflicts
        var hasConflict = await _appointmentRepository.HasConflictingAppointmentAsync(
            dto.StaffId,
            dto.StartTime,
            endTime);

        if (hasConflict)
            throw new InvalidOperationException("This time slot is not available");

        // Validate appointment time is in the future
        if (dto.StartTime <= DateTime.UtcNow)
            throw new InvalidOperationException("Appointment time must be in the future");

        // Create appointment
        var appointment = new Appointment
        {
            TenantId = tenantId,
            SalonId = dto.SalonId,
            CustomerId = customerId,
            StaffId = dto.StaffId,
            ServiceId = dto.ServiceId,
            StartTime = dto.StartTime,
            EndTime = endTime,
            Status = AppointmentStatus.Pending,
            Notes = dto.Notes,
            CustomerNotes = dto.CustomerNotes,
            TotalPrice = service.Price
        };

        await _appointmentRepository.AddAsync(appointment);

        _logger.LogInformation("Appointment created: {AppointmentId} for customer {CustomerId}",
            appointment.Id, customerId);

        // Send confirmation email
        try
        {
            await _notificationService.SendAppointmentConfirmationAsync(appointment.Id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send confirmation email for appointment {AppointmentId}", appointment.Id);
            // Don't fail the appointment creation if email fails
        }

        return MapToDto(appointment, service, staff, salon);
    }

    public async Task<AppointmentDetailsDto?> GetAppointmentDetailsAsync(Guid id, Guid tenantId)
    {
        var appointment = await _appointmentRepository.GetAppointmentWithDetailsAsync(id);

        if (appointment == null || appointment.TenantId != tenantId)
            return null;

        return new AppointmentDetailsDto
        {
            Id = appointment.Id,
            TenantId = appointment.TenantId,
            SalonId = appointment.SalonId,
            SalonName = appointment.Salon.Name,
            CustomerId = appointment.CustomerId,
            CustomerName = $"{appointment.Customer.FirstName} {appointment.Customer.LastName}",
            CustomerEmail = appointment.Customer.Email,
            CustomerPhone = appointment.Customer.Phone,
            StaffId = appointment.StaffId,
            StaffName = $"{appointment.Staff.FirstName} {appointment.Staff.LastName}",
            ServiceId = appointment.ServiceId,
            ServiceName = appointment.Service.Name,
            ServiceDuration = appointment.Service.DurationMinutes,
            ServicePrice = appointment.Service.Price,
            StartTime = appointment.StartTime,
            EndTime = appointment.EndTime,
            Status = appointment.Status,
            Notes = appointment.Notes,
            CustomerNotes = appointment.CustomerNotes,
            CancellationReason = appointment.CancellationReason,
            CancelledAt = appointment.CancelledAt,
            TotalPrice = appointment.TotalPrice,
            DepositPaid = appointment.DepositPaid,
            ReminderSent = appointment.ReminderSent,
            CreatedAt = appointment.CreatedAt,
            Salon = new SalonDto
            {
                Id = appointment.Salon.Id,
                TenantId = appointment.Salon.TenantId,
                Name = appointment.Salon.Name,
                Address = appointment.Salon.Address,
                City = appointment.Salon.City,
                Phone = appointment.Salon.Phone,
                Email = appointment.Salon.Email,
                State = appointment.Salon.State,
                PostalCode = appointment.Salon.PostalCode,
                Latitude = appointment.Salon.Latitude,
                Longitude = appointment.Salon.Longitude,
                BusinessHours = appointment.Salon.BusinessHours,
                IsActive = appointment.Salon.IsActive,
                AverageRating = appointment.Salon.AverageRating,
                ReviewCount = appointment.Salon.ReviewCount,
                CreatedAt = appointment.Salon.CreatedAt
            },
            Service = new ServiceDto
            {
                Id = appointment.Service.Id,
                Name = appointment.Service.Name,
                Description = appointment.Service.Description,
                DurationMinutes = appointment.Service.DurationMinutes,
                Price = appointment.Service.Price,
                ImageUrl = appointment.Service.ImageUrl,
                CategoryId = appointment.Service.CategoryId,
                CategoryName = appointment.Service.Category?.Name,
                IsActive = appointment.Service.IsActive
            },
            Staff = new StaffDto
            {
                Id = appointment.Staff.Id,
                TenantId = appointment.Staff.TenantId,
                SalonId = appointment.Staff.SalonId,
                FirstName = appointment.Staff.FirstName,
                LastName = appointment.Staff.LastName,
                FullName = $"{appointment.Staff.FirstName} {appointment.Staff.LastName}",
                Email = appointment.Staff.Email,
                Phone = appointment.Staff.Phone,
                Bio = appointment.Staff.Bio,
                PhotoUrl = appointment.Staff.PhotoUrl,
                ProfilePictureUrl = appointment.Staff.PhotoUrl,
                AverageRating = appointment.Staff.AverageRating,
                Status = appointment.Staff.Status,
                InvitationStatus = appointment.Staff.InvitationStatus,
                RoleId = appointment.Staff.RoleId,
                RoleName = appointment.Staff.Role?.Name
            },
            Payments = appointment.Payments?.Select(p => new PaymentDto
            {
                Id = p.Id,
                Amount = p.Amount,
                Method = p.Method,
                Status = p.Status
            }).ToList() ?? new()
        };
    }

    public async Task<AppointmentDto> UpdateAppointmentAsync(UpdateAppointmentDto dto, Guid tenantId)
    {
        var appointment = await _appointmentRepository.GetAppointmentWithDetailsAsync(dto.Id);

        if (appointment == null || appointment.TenantId != tenantId)
            throw new InvalidOperationException("Appointment not found");

        if (appointment.Status == AppointmentStatus.Completed ||
            appointment.Status == AppointmentStatus.Cancelled)
            throw new InvalidOperationException("Cannot update completed or cancelled appointments");

        // If changing staff, validate the new staff
        if (dto.StaffId.HasValue && dto.StaffId.Value != appointment.StaffId)
        {
            var staff = await _staffRepository.GetByIdAsync(dto.StaffId.Value);
            if (staff == null || staff.TenantId != tenantId || staff.SalonId != appointment.SalonId)
                throw new InvalidOperationException("Invalid staff selection");

            appointment.StaffId = dto.StaffId.Value;
        }

        // If changing time, check for conflicts
        if (dto.StartTime != appointment.StartTime)
        {
            var endTime = dto.StartTime.AddMinutes(appointment.Service.DurationMinutes);

            var hasConflict = await _appointmentRepository.HasConflictingAppointmentAsync(
                appointment.StaffId,
                dto.StartTime,
                endTime,
                appointment.Id);

            if (hasConflict)
                throw new InvalidOperationException("This time slot is not available");

            appointment.StartTime = dto.StartTime;
            appointment.EndTime = endTime;
        }

        if (dto.Notes != null)
            appointment.Notes = dto.Notes;

        await _appointmentRepository.UpdateAsync(appointment);

        _logger.LogInformation("Appointment updated: {AppointmentId}", appointment.Id);

        return MapToDto(appointment, appointment.Service, appointment.Staff, appointment.Salon);
    }

    public async Task<AppointmentDto> UpdateAppointmentStatusAsync(
        Guid id,
        UpdateAppointmentStatusDto dto,
        Guid tenantId)
    {
        var appointment = await _appointmentRepository.GetAppointmentWithDetailsAsync(id);

        if (appointment == null || appointment.TenantId != tenantId)
            throw new InvalidOperationException("Appointment not found");

        // Validate status transition
        if (!IsValidStatusTransition(appointment.Status, dto.Status))
            throw new InvalidOperationException($"Cannot change status from {appointment.Status} to {dto.Status}");

        appointment.Status = dto.Status;

        if (dto.Status == AppointmentStatus.Cancelled)
        {
            appointment.CancellationReason = dto.CancellationReason;
            appointment.CancelledAt = DateTime.UtcNow;
        }

        await _appointmentRepository.UpdateAsync(appointment);

        _logger.LogInformation("Appointment status updated: {AppointmentId} to {Status}", id, dto.Status);

        // Send status update notification
        try
        {
            await _notificationService.SendStatusUpdateNotificationAsync(id, dto.Status.ToString());
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send status update notification for appointment {AppointmentId}", id);
        }

        return MapToDto(appointment, appointment.Service, appointment.Staff, appointment.Salon);
    }

    public async Task CancelAppointmentAsync(Guid id, string? cancellationReason, Guid tenantId)
    {
        var appointment = await _appointmentRepository.GetByIdAsync(id);

        if (appointment == null || appointment.TenantId != tenantId)
            throw new InvalidOperationException("Appointment not found");

        if (appointment.Status == AppointmentStatus.Completed)
            throw new InvalidOperationException("Cannot cancel completed appointments");

        if (appointment.Status == AppointmentStatus.Cancelled)
            throw new InvalidOperationException("Appointment is already cancelled");

        appointment.Status = AppointmentStatus.Cancelled;
        appointment.CancellationReason = cancellationReason;
        appointment.CancelledAt = DateTime.UtcNow;

        await _appointmentRepository.UpdateAsync(appointment);

        _logger.LogInformation("Appointment cancelled: {AppointmentId}", id);

        // Send cancellation notification
        try
        {
            await _notificationService.SendCancellationNotificationAsync(id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send cancellation notification for appointment {AppointmentId}", id);
        }
    }

    public async Task<AppointmentDto> RescheduleAppointmentAsync(
        Guid id,
        DateTime newStartTime,
        Guid? newStaffId,
        Guid tenantId)
    {
        var appointment = await _appointmentRepository.GetAppointmentWithDetailsAsync(id);

        if (appointment == null || appointment.TenantId != tenantId)
            throw new InvalidOperationException("Appointment not found");

        if (appointment.Status == AppointmentStatus.Completed ||
            appointment.Status == AppointmentStatus.Cancelled)
            throw new InvalidOperationException("Cannot reschedule completed or cancelled appointments");

        var staffId = newStaffId ?? appointment.StaffId;
        var oldStartTime = appointment.StartTime; // Store old time for notification
        var endTime = newStartTime.AddMinutes(appointment.Service.DurationMinutes);

        // Check for conflicts
        var hasConflict = await _appointmentRepository.HasConflictingAppointmentAsync(
            staffId,
            newStartTime,
            endTime,
            appointment.Id);

        if (hasConflict)
            throw new InvalidOperationException("This time slot is not available");

        if (newStaffId.HasValue && newStaffId.Value != appointment.StaffId)
        {
            var staff = await _staffRepository.GetByIdAsync(newStaffId.Value);
            if (staff == null || staff.TenantId != tenantId || staff.SalonId != appointment.SalonId)
                throw new InvalidOperationException("Invalid staff selection");

            appointment.StaffId = newStaffId.Value;
        }

        appointment.StartTime = newStartTime;
        appointment.EndTime = endTime;
        appointment.Status = AppointmentStatus.Pending; // Reset to pending after rescheduling

        await _appointmentRepository.UpdateAsync(appointment);

        // Reload with details
        appointment = await _appointmentRepository.GetAppointmentWithDetailsAsync(id);

        _logger.LogInformation("Appointment rescheduled: {AppointmentId}", id);

        // Send reschedule notification
        try
        {
            await _notificationService.SendRescheduleNotificationAsync(id, oldStartTime);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send reschedule notification for appointment {AppointmentId}", id);
        }

        return MapToDto(appointment!, appointment!.Service, appointment.Staff, appointment.Salon);
    }

    public async Task<IEnumerable<AppointmentDto>> GetCustomerAppointmentsAsync(
        Guid customerId,
        DateTime? startDate = null,
        DateTime? endDate = null,
        AppointmentStatus? status = null)
    {
        var appointments = await _appointmentRepository.GetAppointmentsByCustomerAsync(
            customerId,
            startDate,
            endDate,
            status);

        return appointments.Select(a => MapToDto(a, a.Service, a.Staff, a.Salon)).ToList();
    }

    public async Task<IEnumerable<AppointmentDto>> GetStaffAppointmentsAsync(
        Guid staffId,
        Guid tenantId,
        DateTime? startDate = null,
        DateTime? endDate = null)
    {
        var staff = await _staffRepository.GetByIdAsync(staffId);
        if (staff == null || staff.TenantId != tenantId)
            throw new InvalidOperationException("Staff not found");

        var appointments = await _appointmentRepository.GetAppointmentsByStaffAsync(
            staffId,
            startDate,
            endDate);

        return appointments.Select(a => MapToDto(a, a.Service, a.Staff, a.Salon)).ToList();
    }

    public async Task<IEnumerable<AppointmentDto>> GetSalonAppointmentsAsync(
        Guid salonId,
        Guid tenantId,
        DateTime? startDate = null,
        DateTime? endDate = null)
    {
        var salon = await _salonRepository.GetByIdAsync(salonId);
        if (salon == null || salon.TenantId != tenantId)
            throw new InvalidOperationException("Salon not found");

        var appointments = await _appointmentRepository.GetAppointmentsBySalonAsync(
            salonId,
            startDate,
            endDate);

        return appointments.Select(a => MapToDto(a, a.Service, a.Staff, a.Salon)).ToList();
    }

    public async Task<IEnumerable<AvailableTimeSlotDto>> GetAvailableTimeSlotsAsync(
        Guid staffId,
        DateTime date,
        int serviceDurationMinutes)
    {
        var slots = await _appointmentRepository.GetAvailableTimeSlotsAsync(
            staffId,
            date,
            serviceDurationMinutes);

        return slots.Select(s => new AvailableTimeSlotDto
        {
            Date = s.Date,
            StartTime = s.TimeOfDay,
            EndTime = s.AddMinutes(serviceDurationMinutes).TimeOfDay,
            DurationMinutes = serviceDurationMinutes
        }).ToList();
    }

    public async Task<bool> HasConflictingAppointmentAsync(
        Guid staffId,
        DateTime startTime,
        DateTime endTime,
        Guid? excludeAppointmentId = null)
    {
        return await _appointmentRepository.HasConflictingAppointmentAsync(
            staffId,
            startTime,
            endTime,
            excludeAppointmentId);
    }

    private AppointmentDto MapToDto(Appointment appointment, Service service, Staff staff, Salon salon)
    {
        return new AppointmentDto
        {
            Id = appointment.Id,
            TenantId = appointment.TenantId,
            SalonId = appointment.SalonId,
            SalonName = salon.Name,
            CustomerId = appointment.CustomerId,
            CustomerName = $"{appointment.Customer?.FirstName} {appointment.Customer?.LastName}",
            CustomerEmail = appointment.Customer?.Email ?? "",
            CustomerPhone = appointment.Customer?.Phone,
            StaffId = appointment.StaffId,
            StaffName = $"{staff.FirstName} {staff.LastName}",
            ServiceId = appointment.ServiceId,
            ServiceName = service.Name,
            ServiceDuration = service.DurationMinutes,
            ServicePrice = service.Price,
            StartTime = appointment.StartTime,
            EndTime = appointment.EndTime,
            Status = appointment.Status,
            Notes = appointment.Notes,
            CustomerNotes = appointment.CustomerNotes,
            CancellationReason = appointment.CancellationReason,
            CancelledAt = appointment.CancelledAt,
            TotalPrice = appointment.TotalPrice,
            DepositPaid = appointment.DepositPaid,
            ReminderSent = appointment.ReminderSent,
            CreatedAt = appointment.CreatedAt
        };
    }

    private bool IsValidStatusTransition(AppointmentStatus currentStatus, AppointmentStatus newStatus)
    {
        // Define valid status transitions
        return (currentStatus, newStatus) switch
        {
            (AppointmentStatus.Pending, AppointmentStatus.Confirmed) => true,
            (AppointmentStatus.Pending, AppointmentStatus.Cancelled) => true,
            (AppointmentStatus.Confirmed, AppointmentStatus.CheckedIn) => true,
            (AppointmentStatus.Confirmed, AppointmentStatus.Cancelled) => true,
            (AppointmentStatus.Confirmed, AppointmentStatus.NoShow) => true,
            (AppointmentStatus.CheckedIn, AppointmentStatus.InProgress) => true,
            (AppointmentStatus.CheckedIn, AppointmentStatus.Cancelled) => true,
            (AppointmentStatus.InProgress, AppointmentStatus.Completed) => true,
            _ => false
        };
    }
}
